Overrides MailChimp settings for 26+ Club.

sergiitk opened the pull request:
#### What's this PR do?
- Introduces 3 new setting overrides for 26+ Club:
  1. `dosomething_signup_mailchimp_26plusclub_list_id`: Overrides `mailchimp_list_id`
  2. `dosomething_signup_mailchimp_26plusclub_grouping_id_register`: Overrides `mailchimp_grouping_id` for registrations
  3. `dosomething_signup_mailchimp_26plusclub_group_name_register`: Overrides `mailchimp_group_name` for registrations
- Extends and imporves [Third Party Opt-Ins](http://dev.dosomething.org:8888/admin/config/dosomething/opt_in) configuration interface:
  ![image](https://cloud.githubusercontent.com/assets/672669/6757629/1ac7ccba-cf3b-11e4-9c76-798122a31bd2.png)
- Turns off MBP for 26+ Club signups
- Overrides settings for 26+ Club registrations
#### Any background context you want to provide?

New user payload before:

``` php
array (size=11)
  'activity' => string 'user_register' (length=13)
  'email' => string 'sergii+test-032015-7@dosomething.org' (length=36)
  'uid' => string '1704217' (length=7)
  'email_template' => string 'mb-user-register-US' (length=19)
  'merge_vars' => 
    array (size=2)
      'MEMBER_COUNT' => string '3 million' (length=9)
      'FNAME' => string 'Sergii' (length=6)
  'mailchimp_list_id' => string '* hidden *' (length=10)
  'birthdate' => string '580780800' (length=9)
  'subscribed' => int 1
  'email_tags' => 
    array (size=1)
      0 => string 'drupal_user_register' (length=20)
  'activity_timestamp' => int 1426871199
  'application_id' => string 'US' (length=2)
```

New user payload after:

``` php
array (size=13)
  'activity' => string 'user_register' (length=13)
  'email' => string 'sergii+test-032015-7@dosomething.org' (length=36)
  'uid' => string '1704233' (length=7)
  'email_template' => string 'mb-user-register-US' (length=19)
  'merge_vars' => 
    array (size=2)
      'MEMBER_COUNT' => string '3 million' (length=9)
      'FNAME' => string 'Sergii' (length=6)
  'mailchimp_list_id' => string '* hidden-2 *' (length=10)
  'birthdate' => string '580780800' (length=9)
  'subscribed' => int 1
  'mailchimp_grouping_id' => string '9613' (length=4)
  'mailchimp_group_name' => string 'Web Registrations' (length=17)
  'email_tags' => 
    array (size=1)
      0 => string 'drupal_user_register' (length=20)
  'activity_timestamp' => int 1426873547
  'application_id' => string 'US' (length=2)
```
#### What are the relevant tickets?

A part of #4058

sergiitk committed:
Overrides MailChimp settings for 26+ Club.

aaronschachter commented:
It looks like this may close #1483 also...

sergiitk commented:
@aaronschachter not sure. Now it overrides grouping id and group name only for 26+. Should it provide this for normal regs also?

aaronschachter commented:
Oh, ok, nevermind (out of scope for what you're working on)

sergiitk commented:
@aaronschachter I can easily implement this after.

deezone commented:
Please explain what "Turns off MBP for 26+ Club signups" in the PR description means? I'm under the impression nothing is "turned off" but rather adjustments to:

```
  'mailchimp_list_id'
  'mailchimp_grouping_id' 
  'mailchimp_group_name'
```

being made? Perhaps this is in reference to international sites where only `application_id = US` is being applied to the 26+ functionality?

sergiitk commented:
@DeeZone This PR adjusts this values for new registrations, but disables requests to MBP for campaign signups, as described in #4058:

> Campaign Signup:  Not be added to the DoSomething Members list in MailChimp.

And I don't see any reference to international sites in #4058. 
Did I missed something?

deezone commented on lib/modules/dosomething/dosomething_signup/dosomething_signup.module:
```diff
@@ -465,6 +465,12 @@ function dosomething_signup_mbp_request($account, $node, $opt_in) {
   }
+
+  // Exclude 26+ Club.
+  if (dosomething_user_is_old_person($account)) {
```
Ah, this it the exclude you're talking about. Looks good.

deezone commented:
And yes, the 26+ logic for international sites is a different issue so my comment was out of context.

Everything looks good :+1:

deezone commented:
@sergii-tkachenko This functionality should **only** be applied on the US site. All other affiliates should not be affected by user age. Perhaps by not setting the 26+ values in the affiliate sites this functionality will be disabled?

Does this address your question?

sergiitk commented:
@DeeZone But what happens on campaign signups? This functionality turns it MBP requests for signups completely. Is this right for international?

deezone commented:
In all cases **for affiliates**, the users age should not affect the current functionality. This PR should only be applied to the US site functionality or rather, if the 26+ Club settings are set. All 26+ Club functionality should be disabled when there are no values in the new admin form fields for 26+ Club.

sergiitk commented:
Ok, changing this PR according to latest requirements.

sergiitk committed:
MailChimp List ID: fallback to general if 26+ Club value isn't set.

sergiitk committed:
Various changes to 26+ Club.

- Add setting to disable 26+ Club
- Makes 26+ Club overrides optional
- Provides a function to determine if user is a member of the club
- Overrides MBP requests only when the club is enabled and only
  variables that are set

sergiitk commented:
Multiple changes:
- Add setting to disable 26+ Club
  ![image](https://cloud.githubusercontent.com/assets/672669/6808151/27fdbf02-d25b-11e4-8975-0656fea7bc3d.png)
- Makes 26+ Club overrides optional
- Provides a function to determine if user is a member of the club
- Overrides MBP requests only when the club is enabled and only
  variables that are set

@DeeZone please re-review.

deezone commented:
@sergii-tkachenko still working on testing. I think I might have found a bug not due to this PR but related.
- [x] Add setting to disable 26+ Club

I understand why you've added the checkbox to make it clear that the values need to be filled out. 

What is supposed to happen if the checkbox is enabled but there are no values in the form fields? Perhaps:
- Overrides MBP requests only when the club is enabled and only variables that are set.
  If the variables are not set then the checkbox is just a UI element that has no effect?
- [x] Provides a function to determine if user is a member of the club

sergiitk commented:
@DeeZone I the box is enabled, but variables are not set:
- Campaign signups will not be sent for 26+ Club users
- Registrations will be sent with normal (not overriden) MailChimp values.

deezone commented:
@sergii-tkachenko Fantastic - can you document this functionality in the new form elements description?

sergiitk commented:
Sure!

sergiitk committed:
Adds description for Enable 26+ Club checkbox.

sergii-tkachenko commented:
@DeeZone Something like this?
![image](https://cloud.githubusercontent.com/assets/672669/6851521/582b37ca-d3e8-11e4-87c3-0682179505d3.png)

sergii-tkachenko commented:
@DeeZone Something like this?
![image](https://cloud.githubusercontent.com/assets/672669/6851572/a07110d6-d3e8-11e4-8583-809caf8089bc.png)

deezone commented:
- [x] User registration only - under 26+, setting not enabled:
  - [x] **mobileCommonsQueue**: 
    - No opt_in ID when mobile number not included in registration.
    - `"mobile";s:10:"1234560984";s:17:"mc_opt_in_path_id";s:6:"164905"`
  - [x] **userRegistrationQueue**: `s:17:"mailchimp_list_id";s:10:"f2fab1dfd4";`
- [ ] User registration only - over 26+, setting not enabled:
  - [ ] **mobileCommonsQueue**: 
    - **??** No opt_in ID when **mobile number included** in registration.
  - [x] **userRegistrationQueue**: `s:17:"mailchimp_list_id";s:10:"f2fab1dfd4";`
- [x] User registration only - under 26+, setting enabled (checked), no values:
  - [x] **mobileCommonsQueue**:  `"mobile";s:10:"1234560982";s:17:"mc_opt_in_path_id";s:6:"164905"`
  - [x] **userRegistrationQueue**: `s:17:"mailchimp_list_id";s:10:"f2fab1dfd4"`
- [x] User registration only - over 26+, setting enabled (checked), no values:
  - [x] **mobileCommonsQueue**: 
    - No opt_in ID when **mobile number included** in registration.
  - [x] **userRegistrationQueue**: `"mailchimp_list_id";s:10:"f2fab1dfd4"`
- [x] User registration only - under 26+, setting enabled (checked), **with values**:
  - [x] **mobileCommonsQueue**:  `"s:6:"mobile";s:10:"1234560980";s:17:"mc_opt_in_path_id";s:6:"164905"`
  - [x] **userRegistrationQueue**: `"mailchimp_list_id";s:10:"f2fab1dfd4"`
- [x] User registration only - over 26+, setting enabled (checked), **with values**:
  - [x] **mobileCommonsQueue**: 
    - No opt_in ID when **mobile number included** in registration.
  - [x] **userRegistrationQueue**: `"mailchimp_list_id";s:22:"mailchimp_list_id_1234";"mailchimp_grouping_id";s:26:"mailchimp_grouping_id_1234";s:20:"mailchimp_group_name";s:25:"mailchimp_group_name_1234"`
    :"mailchimp_list_id";s:22:"mailchimp_list_id_1234";

This covers all the user registration only test. It appears there's an issue with:
- [ ] User registration only - over 26+, setting not enabled:
  - [ ] **mobileCommonsQueue**: 
    - **??** No opt_in ID when **mobile number included** in registration.
      The mobile number should still be added to Mobile Commons when the 26+ setting is disabled.

deezone commented:
- [x] User registration and campaign signup - under 26+, setting not enabled:
  - [x] **mobileCommonsQueue**: 
    - [x] No opt_in ID when mobile number not included in registration.
    - [x] opt_in for registration: `s:6:"mobile";s:10:"1234567899";s:17:"mc_opt_in_path_id";s:6:"167401";`
    - [x] opt_in for Staff Pick campaign: `mobile";s:10:"1234560911";s:17:"mc_opt_in_path_id";s:6:"173831"`
  - [x] **userRegistrationQueue**: `s:17:"mailchimp_list_id";s:10:"f2fab1dfd4"`
  - [x] **mailchimpCampaignSignupQueue**: Non Staff Pick, MailChimp details excluded. Details included in Staff Pick: `s:17:"mailchimp_list_id";s:10:"f2fab1dfd4";s:21:"mailchimp_grouping_id";s:16:"grouping_id_1234";s:20:"mailchimp_group_name";s:14:"grup_name_1234";`
- [x] User registration and campaign signup - under 26+, **26+ setting enabled**, empty field values:
  - [x] **mailchimpCampaignSignupQueue**: `s:17:"mailchimp_list_id";s:10:"f2fab1dfd4";s:21:"mailchimp_grouping_id";s:16:"grouping_id_1234";s:20:"mailchimp_group_name";s:14:"grup_name_1234";`
- [x] User registration and campaign signup - **over** 26+, **26+ setting enabled**:
  - [x] **mobileCommonsQueue**: No opt_in value
  - [ ] **userRegistrationQueue**: `s:17:"mailchimp_list_id";s:10:"f2fab1dfd4";` !!! - Missing MailChimp details
- [x] User registration and campaign signup - **over** 26+, **26+ setting enabled**, empty field values:
  - [x] **mobileCommonsQueue**: No opt_in value
  - [x] **userRegistrationQueue**: `s:17:"mailchimp_list_id";s:10:"f2fab1dfd4"` : defaults to base MailChimp list

deezone commented:
@sergii-tkachenko We need to talk about these test results, it appears 26+ users are not producing the expected MailChimp values in the `userRegistrationQueue`. See **over 26+, 26+ setting enabled** entry above.

deezone commented:
@sergii-tkachenko To be more specific, in the case of a user who's over 26+ and the 26+ setting is enabled and the user registers for a campaign. The `mailchimp_list_id` should be excluded from the `activity: campaign_signup` message payload.

deezone commented:
@sergii-tkachenko And further, the issue **User registration only - over 26+, setting not enabled**.
- mobileCommonsQueue: ?? No opt_in ID when mobile number included in registration. 

Please test and take note that when the **26+** setting is **not** enabled the 26+ user is excluded from Mobile Commons registrations by not sending `mobile` and `opt_in` values in message.

sergiitk committed:
Ignore MailChimp on registrations only when 26+ Club is enabled.

sergiitk commented:
@DeeZone Please retest with the last change.

deezone commented:
Tested with @sergii-tkachenko in all combos - looks great :+1:
